### PR TITLE
Add `np.broadcast_to(scalar_array, ())`

### PR DESCRIPTION
**Pull request by guilhermeleobas**
This PR adds a special case where the input to `np.broadcast_to` is a scalar array + empty shape.

### CODE REVIEW

File: numba/np/arrayobj.py
Comment (kc611) [inline numba/np/arrayobj.py:1356]: I have a question here. What's the difference between `parent` as an attribute and `parent` as a type within the datamodel ?

File: numba/np/arrayobj.py
Comment (kc611) [inline numba/np/arrayobj.py:1412]: This branch could use a simple check to see if it's actually an empty tuple and not something like `(1, 'x')`, the current logic accepts that shape which is invalid.

File: numba/np/arrayobj.py
Comment (kc611) [inline numba/np/arrayobj.py:1418]: Minor nitpick ? 
```suggestion
            msg = 'Cannot broadcast a non-scalar to a scalar array'
```

Comment (guilhermeleobas) [reply numba/np/arrayobj.py:1418]: Fixed

File: numba/np/arrayobj.py
Comment (guilhermeleobas) [reply numba/np/arrayobj.py:1412]: Thanks, fixed!

File: numba/np/arrayobj.py
Comment (guilhermeleobas) [reply numba/np/arrayobj.py:1356]: `parent` as an attribute is a member of the ndarray struct, whereas the datamodel says how `parent` should be typed. See:
https://github.com/numba/numba/blob/2e9b58cfce0391b968ab8c7c84393cc12c4d1bfb/numba/core/datamodel/models.py#L876

**Review by kc611 — APPROVED** [review]
The test failure looks unrelated, rest LGTM!

**Comment by esc** [comment]
> The test failure looks unrelated, rest LGTM!

Yes, this should be fixed with #8354 -- but you may need to merge in `main` b/c RTD will build a PR against the branch head and not the branch merge, so the build doesn't include any fixes from `main` (unlike Azure).

**Review by esc — CHANGES_REQUESTED** [review]
Some nitpicks I found.

File: numba/np/arrayobj.py
Comment (guilhermeleobas) [reply numba/np/arrayobj.py]: Yes! Removed

File: numba/np/arrayobj.py
Comment (guilhermeleobas) [reply numba/np/arrayobj.py]: Removed

File: numba/np/arrayobj.py
Comment (stuartarchibald) [inline numba/np/arrayobj.py:1411]: This:
```python
from numba import njit
import numpy as np

@njit
def foo():
    return np.broadcast_to(123, ())

print(foo())
```
produces:
```
    if array.ndim == 0:
AttributeError: 'Integer' object has no attribute 'ndim'
```

**Comment by guilhermeleobas** [comment]
CI failure is not related to this PR.

**Comment by stuartarchibald** [comment]
> CI failure is not related to this PR.

Failed test was: `ERROR: test_np_MachAr_deprecation_np122 (numba.tests.test_np_functions.TestNPMachineParameters)`. It indeed seems unrelated.

**Comment by esc** [comment]
Running on the Anaconda Internal Build farm as `numba_smoketest_cpu_yaml_138`.

File: numba/np/arrayobj.py
Comment (stuartarchibald) [inline numba/np/arrayobj.py:1413]: Perhaps add a comment here saying that `array` must be supported by `type_can_asarray` as checked above. I just spent a few minutes trying to work out why inputs such as `'hello'` are commented in the test below, it's because whilst they are permitted scalar types, they fail the `type_can_asarray` check so cannot reach here.

This also lead to the discovery of this bug:
```python
from numba import njit
import numpy as np


@njit
def foo():
    return np.broadcast_to('a', (2, 3))

print(foo())
```
During type inference, for the non-literal type attempt, this function correctly fails to accept `'a'` as an input (as it's a `unicode_type`). For the subsequent literal type attempt, the types are: `(Literal[str](a), Tuple(Literal[int](2), Literal[int](3)))`. The literal string 'a', is correctly handled and accepted, but the shape tuple is now a `types.Tuple` of literal integers (i.e. heterogeneous), and this is not handled correctly and leads to calling the `else:` branch fall through below.

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the update @guilhermeleobas. There's a couple of issues to resolve else looks good.

**Comment by esc** [comment]
> Running on the Anaconda Internal Build farm as `numba_smoketest_cpu_yaml_138`.

This was all green 🍏 -- but probably will need another review after @stuartarchibald 's comments have been addressed.

File: numba/np/arrayobj.py
Comment (guilhermeleobas) [reply numba/np/arrayobj.py:1424]: Hi Stuart, I've addressed the issue.

**Comment by esc** [comment]
`Build numba_smoketest_cpu_yaml_141 has started`

**Comment by guilhermeleobas** [comment]
@esc, can you please re-run this PR in the build farm?

**Comment by esc** [comment]
> `Build numba_smoketest_cpu_yaml_141 has started`

This was fine, again! Now waiting on a final sign-off from @stuartarchibald .

**Review by stuartarchibald — APPROVED** [review]
Thanks for the patch and fixes!

**Pull request merged**
